### DISCUSSION TITLE
How to exclude values

### DISCUSSION THREAD

**Discussion by cdavila** [mapsforge/mapsforge#1500] [Q&A] [answered]
According to [Rendertheme doc](https://github.com/mapsforge/vtm/blob/master/docs/Rendertheme.md), values can be excluded in rules using a minus sign within the "v" string. I have tried several ways of using the minus sigh without success. For example, if I want to apply a rule to all water=* objects but reservoirs, I try.
`<rule e="way" k="natural" v="water" closed="yes" zoom-min="9">`
`    <rule e="way" k="water" v="reservoir">`
`        <area fill="#3399FF" />`
`    </rule>`
`    <rule e="way" k="water" v="*|-reservoir">`
`        <area fill="#005500" />`
`    </rule>`
`</rule>`
What's the right way to write minus sign in rules? Can anyone provide a working example?

**Comment by devemux86** [comment]
Your link is the VTM render theme documentation.

Mapsforge render theme documentation is [here](https://github.com/mapsforge/mapsforge/blob/master/docs/Rendertheme.md).
This operator is not implemented in Mapsforge.

**Comment by ludwigb** [comment]
Quickly scanning through that, there is no exclusion operator in the
mapsforge renderthemes.

But maybe it would work if you add an element that has the display property
set to 'never'. Worth a try.

**Comment by MotoUKRider** [comment]
Hello @ludwigb I had a similar experience with the [Motorider theme](https://github.com/devemux86/cruiser/discussions/612).

I used the ~ symbol but my expectations of its method were not correct!  I wanted to draw "oneway" arrows on a road that was *not* an off-road "track" - because we have an off-road layer that users turn on/off *and* I had found a very creative track in Scotland that had been set-up as "oneway", some kind of motocross course, I think. My workaround was to define *all* roads except "track".

Expected:
`<rule e="way" k="highway" v="~track" zoom-min="15">`

Actual:
`<rule e="way" k="highway" v="motorway|motorway_link|primary|primary_link|secondary|secondary_link|tertiary|tertiary_link|service|unclassified|residential|living_street|construction|road" zoom-min="15">`

So yes, I think a simple way of excluding a tag (or tags) would be useful.

**Reply by devemux86** [reply]
In VTM it works like this:
```xml
 <!-- match tunnel-tag that are not 'no' or 'false' -->
 <m k="tunnel" v="-|no|false">
```
I will see how easy it is to implement in Mapsforge as well.

**Comment by devemux86** [comment, marked as answer]
I implemented the exclusive / except negation rule:
- https://github.com/mapsforge/mapsforge/pull/1524

**Comment by cdavila** [comment]
Great!!
Thank you very much
